sklearn/impute.py

```diff
@@ -115,16 +141,53 @@ class SimpleImputer(BaseEstimator, TransformerMixin):
     Notes
     -----
     Columns which only contained missing values at `fit` are discarded upon
-    `transform`.
+    `transform` if strategy is not "constant".
 
     """
-    def __init__(self, missing_values="NaN", strategy="mean",
-                 verbose=0, copy=True):
+    def __init__(self, missing_values=np.nan, strategy="mean",
+                 fill_value=None, verbose=0, copy=True):
         self.missing_values = missing_values
         self.strategy = strategy
+        self.fill_value = fill_value
         self.verbose = verbose
         self.copy = copy
 
+    def _validate_input(self, X):
+        allowed_strategies = ["mean", "median", "most_frequent", "constant"]
+        if self.strategy not in allowed_strategies:
+            raise ValueError("Can only use these strategies: {0} "
+                             " got strategy={1}".format(allowed_strategies,
+                                                        self.strategy))
+
+        if self.strategy in ("most_frequent", "constant"):
+            dtype = None
+        else:
+            dtype = FLOAT_DTYPES
+
+        if not is_scalar_nan(self.missing_values):
+            force_all_finite = True
+        else:
+            force_all_finite = "allow-nan"
+
+        try:
+            X = check_array(X, accept_sparse='csc', dtype=dtype,
+                            force_all_finite=force_all_finite, copy=self.copy)
+        except ValueError as ve:
+            if "could not convert" in str(ve):
+                raise TypeError("Cannot use {0} strategy with non-numeric "
+                                "data. Received datatype :{1}."
+                                "".format(self.strategy, X.dtype.kind))
+            else:
+                raise ve
+
+        if X.dtype.kind not in ("i", "u", "f", "O"):
+            raise TypeError("The SimpleImputer does not support this datatype"
+                            " ({0}). Please provide either numeric data or"
+                            " categorical data represented by integer or "
+                            "object datatypes.".format(X.dtype))
```

Comment:
I am wondering: do we usually raise `TypeError` or `ValueError` when the object type is valid, but the `dtype` is not?

Also, the message is a bit confusing: one could get the impression that float data is not supported.

"""
SimpleImputer does not work on data with dtype {0}. Please provide either a numeric array (with a floating point or integer dtype) or categorical data represented either as an array with integer dtype or an array of string values with an object dtype.
"""

Better be explicit in error message. Verbosity is not a problem here.

Reply:
numpy raises TypeError for bad dtypes. E.g. `np.log(np.asarray(['hello']))`

Reply:
numpy is not very consistent:

```python
>>> np.array(['a', 'b'], dtype=object).astype(np.uint8)
Traceback (most recent call last):
  File "<ipython-input-6-3200a6709d46>", line 1, in <module>
    np.array(['a', 'b'], dtype=object).astype(np.uint8)
ValueError: invalid literal for int() with base 10: 'a'
```

It's true that if you consider the individual scalar operation the TypeError raised by a ufunc might make sense. If you consider that the array as a whole is invalid, then ValueError makes more sense. I think both are fine in this case.